src/base/QXmppExternalServiceDiscoveryIq.cpp

```diff
@@ -0,0 +1,484 @@
+{
+    d->username = username;
+}
+
+///
+/// Returns true if the element is a valid external service and can be parsed.
+///
+bool QXmppExternalService::isExternalService(const QDomElement &element)
+{
+    if (element.tagName() != "service") {
+        return false;
+    }
+
+    return element.hasAttribute("host") && !element.attribute("host").isEmpty() &&
+        element.hasAttribute("type") && !element.attribute("type").isEmpty();
+}
+
+///
+/// Parses given DOM element as an external service.
+///
+void QXmppExternalService::parse(const QDomElement &el)
+{
+    const QDomNamedNodeMap &attributes { el.attributes() };
+
+    setHost(el.attribute("host"));
+    setType(el.attribute("type"));
+
+    if (attributes.contains("action")) {
+        const std::optional<Action> action { actionFromString(el.attribute("action")) };
+
+        if (action.has_value()) {
+            setAction(action.value());
+        }
+    }
+
+    if (attributes.contains("expires")) {
+        const QString &expiresAttribute { el.attribute("expires") };
+
+        QDateTime expires;
+        Qt::DateFormat dateFormat {
+			expiresAttribute.contains(QRegularExpression("\\.[0-9]{3}"))
+                ? Qt::DateFormat::ISODateWithMs
+                : Qt::DateFormat::ISODate
+        };
+
+        expires = QDateTime::fromString(expiresAttribute, dateFormat);
+        expires.setTimeZone(QTimeZone::utc());
+
+        setExpires(expires);
+    }
+
+    if (attributes.contains("name")) {
+        setName(el.attribute("name"));
+    }
+
+    if (attributes.contains("password")) {
+        setPassword(el.attribute("password"));
+    }
+
+    if (attributes.contains("port")) {
+        setPort(el.attribute("port").toInt());
+    }
+
+    if (attributes.contains("restricted")) {
+        setRestricted(el.attribute("restricted") == "true" ? true : false);
+    }
+
+    if (attributes.contains("transport")) {
+        const std::optional<Transport> transport { transportFromString(el.attribute("transport")) };
+
+        if (transport.has_value()) {
+            setTransport(transport.value());
+        }
+    }
+
+    if (attributes.contains("username")) {
+        setUsername(el.attribute("username"));
+    }
+}
+
+///
+/// \brief QXmppExternalService::toXml
+///
+/// Translates the external service to XML using the provided XML stream writer.
+///
+/// \param writer
+///
+
+void QXmppExternalService::toXml(QXmlStreamWriter *writer) const
+{
+    writer->writeStartElement("service");
+    helperToXmlAddAttribute(writer, "host", host());
+    helperToXmlAddAttribute(writer, "type", type());
+
+    if (action().has_value()) {
+        const std::optional<Action> action { this->action() };
+
+        if (action.has_value()) {
+            helperToXmlAddAttribute(writer, "action", actionToString(action.value()));
+        }
+    }
+
+    if (expires().has_value()) {
+        helperToXmlAddAttribute(writer, "expires", expires().value().toString(Qt::ISODateWithMs));
```

Comment:
```suggestion
        helperToXmlAddAttribute(writer, "expires", d->expires->toString(Qt::ISODateWithMs));
```
`std::optional` supports the `->` operator :)